compiler/src/dotty/tools/dotc/escapecheck/EscapeCheck.scala

```diff
@@ -0,0 +1,97 @@
+package dotty.tools.dotc
+package escapecheck
+
+import core._
+import Types._
+import Contexts.Context
+import Symbols._
+import Decorators._
+import Phases._
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.ast.tpd
+
+
+class EscapeCheck extends Phase {
+  import tpd._
+
+  override def run(implicit ctx: Context): Unit = {
+    val curTree = ctx.compilationUnit.tpdTree
+
+    traverser((),curTree)(ctx.fresh.setLocalMode(firstClass))
+  }
+
+  override def phaseName = "escapeCheck"
+
+  val traverser : TreeTraverser = new TreeTraverser {
+    override def traverse(tree: Tree)(implicit ctx: Context): Unit = tree match {
+      case _ : Ident =>
+
+        if (isSecondClass(tree)) {
+          if (ctx.localMode.eq(firstClass))
+            ctx.error(em"Found local ${tree.symbol} in first class position.", tree.pos)
+          else {
+            for (s <- ctx.boundary) {
+              if (!tree.symbol.denot.isContainedIn(s) && !isSecondClass(s))
+                ctx.error(em"Found local ${tree.symbol} inside $s.", tree.pos)
+            }
+          }
+
+        }
+
+
+      case Block(stats @ (fn: DefDef) :: Nil, Closure(_, fnRef, tpt)) if fnRef.symbol == fn.symbol =>
+        print("Foo")
+
+      case tree : ValDef =>
+        traverseWithLocalMode(tree.rhs, getClassiness(tree))
+
+      case tree : DefDef =>
+        traverse(tree.rhs)(ctx.fresh.setLocalMode(firstClass).addBoundary(tree.symbol))
+
+      case Assign(lhs, rhs) =>
+        traverseWithLocalMode(rhs, getClassiness(lhs))
+
+      case Apply(fn, args) =>
+        traverseWithLocalMode(fn, secondClass)
+
+        val isLocalMethod = fn.tpe match {
+          case tpe : CachedTermRef => tpe.underlying.isLocalMethod
+          case tpe => tpe.isLocalMethod
+        }
+
+        if (isLocalMethod)
+          args.foreach(traverseWithLocalMode(_, secondClass))
+        else
+          args.foreach(traverseWithLocalMode(_, firstClass))
+
+
+      case Block(stats, expr) =>
+        stats.foreach(traverseWithLocalMode(_, secondClass))
```

Comment:
Why do you traverse with `secondClass` and not with the current classiness?

Reply:
```scala
def foo(local x: Int) = {
  x
  42
}
```

Reply:
Yes, you're right.